### ISSUE TITLE
Precision issues with GPS format in astropy.time.Time

### ISSUE THREAD

**Issue by duncanmmacleod** [astropy/astropy#6190]
There is a precision problem somewhere in the implementation of `astropy.time.Time`, at least relative to GPS times. Running on macOS 10.12.5 with python 2.7.13 and astropy 1.3.3:

```python
>>> from astropy.time import Time
>>> Time(1126259462, format='gps').gps == 1126259462
False
```

This presumably is a result of the way the time is stored internally as two JD values, but I have no idea how to debug that code...

**Comment by astrofrog** [comment]
The fractional precision is basically the 64-bit floating point accuracy:

```
>>> Time(1126259462, format='gps').gps - 1126259462
2.384185791015625e-07
>>> np.spacing(1126259462.)
2.384185791015625e-07
```

Do we expect this to be any better? I don't think == should ever be used to compare floats, right?

**Comment by duncanmmacleod** [comment]
@astrofrog, I'm happy to be told I'm using the code wrong, most of the time in LIGO GPS times are stored as two ints representing integer seconds and nanoseconds, which preserves precision up to the nanosecond, I was hoping the underlying ``jd1, jd2`` storage might be leveraged to provide the same thing.

**Comment by astrofrog** [comment]
I'm out of my depth here - I guess I was more thinking that whatever happens internally, at the end of the day you are extracting a single float out, and the accuracy of that float is currently almost as good as it could be (you could not represent nanosecond accuracy with a single float)

**Comment by mhvk** [comment]
@duncanmmacleod - the problem is indeed the conversion to a single double-precision float that you get by doing `.gps`. The actual internal precision is much better:
```
In [6]: (Time(1126259462, format='gps') - Time(1126259452, format='gps')).sec -10.
Out[6]: 3.161915174132446e-12
```
Note that in `Time`, storage is effectively integer days and fraction, so the precision is good to
```
In [8]: (np.spacing(1.) * u.day).to(u.ns)
Out[8]: <Quantity 0.019184653865522705 ns>
```
i.e., you can always recover a LIGO timestamp.

In the end, the main recommendation is to use `Time` consistently, only converting to any quantities relative to a reference time that is relatively nearby.

**Comment by mwcraig** [comment]
Might this make a good example for the documentation?

**Comment by astrofrog** [comment]
Just out of curiosity, is there any way we could allow the user to return Decimal values instead of floats if they want?

**Comment by mhvk** [comment]
Possibly for the representation, but all transformations get done with doubles, so I'm not sure what it would solve.

**Comment by dmalishev** [comment]
There is probably similar issue with mjd - to- mjd conversion via gps seconds, e.g.
```
>>> t = Time(57827.0 , format='mjd' )
>>> t1 = Time(t.gps , format='gps' )
>>> t1.mjd
57827.00042824074
```

**Comment by mhvk** [comment]
@dmalishev - in your example, what goes wrong is that if you use `format='gps'`, that implies `scale='tai'`, while your `t` has the default scale, `utc`. Taking that into account, the times actually are equal:
```
t.scale
# 'utc'
t1.scale
# 'tai'
t1-t
# <TimeDelta object: scale='tai' format='jd' value=0.0>
t1.utc.mjd
# 57827.0
```

**Comment by mhvk** [comment]
p.s. I guess this is one disadvantage to having decided that the user did not have to explicitly set the scale, and let its default be `utc`...

**Comment by taldcroft** [comment]
This corner case is a small price to pay.. :smile:

The original problem does make me think again about making `gps` into a proper scale instead of a format.  If that were the case then at least @duncanmmacleod would have access to `jd1` and `jd2`  which play the same role as the LIGO double-int representation.  I seem to recall being on the negative side of doing `gps` as a scale, but I don't recall the actual reason.

**Comment by dmalishev** [comment]
Oh, I see. Thanks a lot!

**Comment by mhvk** [comment]
I agree that we probably should have made `gps` a scale - it would also be the cleanest way to address that it has (very small) variations in its offset from tai (#2134). But it is tricky: people may by now well count on `t.gps` to give a float64 array...  So, we might have to call the scale something else. Not completely sure that is worth it...

**Comment by astrofrog** [comment]
I think @gahjelle was also in favor of making gps a scale based on discussions I had with him in the past. Is it too late?

**Comment by taldcroft** [comment]
Agreed on just using a different name even if it is not ideal.

**Comment by astrofrog** [comment]
Note that we could still support ``scale='gps'`` since that would be unambiguous, so I guess it's just the 'quick' access via the ``gps`` attribute that would be confusing.

**Comment by taldcroft** [comment]
As reminder, there is useful discussion also in #1879.

The trick about adding a time scale is that all current time scale transformations are implemented by ERFA code.  Thus adding a time scale will require new infrastructure in astropy.time.  It is probably not that hard, but it won't be cookie-cutter like adding a new format.

And of course getting the ~few ns accuracy for `gps` discussed in #2134 (instead of just a fixed offset) would probably be moderately difficult as it appears to require use an online bulletin with corrections.

**Comment by gahjelle** [comment]
Thanks @astrofrog for making me aware of this discussion.

Indeed, we need to use `gps` as a scale in our project (software for amongst other things calculating GPS-satellite orbits). Our current solution is that we have subclassed `astropy.time.Time` and added the `gps`-scale as well as a few other things.

Unfortunately, the structure of `astropy.time.Time` is not ideal for such sub-classing and adding scales. In particular, most of the information about scales is kept in global tuples and dicts in `astropy.time.core` and `astropy.time.formats`. For now, we have simply monkey patched these tuples and dicts. Similarly for the issue @taldcroft raises about all current time scale transformations assumed to be implemented by ERFA: Again our solution has been to monkey patch the ERFA-library.

Obviously, this monkey patching is far from ideal. We have two suggestions for improving this:

1) Simply refactor the `Time`-class so that new scales and formats can be added using some simple hooks. I started looking into this some while ago, but got sidetracked. If there is interest, I can finish this and send over a suggestion?
In this scenario, it is not necessary to actually add `gps` as a scale in core astropy, but it could be a prominent example of how to add your own scales (assuming that there is not a lot of demand for the gps-scale). On the other hand, it should be fairly easy to also add the `gps`-scale as an "official" astropy scale as well.

2) Add the `gps` scale into the current architecture (with any necessary changes). As noted in the earlier discussion, there would be a few small challenges (naming, ERFA, etc), but it should not be too hard. Our monkey patching is about 20 lines of actual code (see https://gist.github.com/gahjelle/1c4e18f0df4e98a8836edcaa6dd1a7a8).

For our project, we are happy with just implementing `gps` at a constant offset from `tai`.

**Comment by taldcroft** [comment]
See #6468.

**Comment by abigailStev** [comment]
Pinging @aarchiba in case you have some insights to contribute here.

**Comment by duncanmmacleod** [comment]
we're discussing this at PiA 2018, so just to post a slightly more concrete example of what the original problem is (that isn't just float64 precision issues):

```python
>>> Time(1234567890, .123456789, format='gps', scale='utc').gps
1234567890.1234567
```

**Comment by duncanmmacleod** [comment]
@mhvk, I would be happy to try and hack at this, but have no idea where to begin. Can you point me at an existing scale definition that would be easy to use as a template?

**Comment by nikkelj** [comment]
I agree, it should be modeled as a constant offset, TAI+19 seconds (the number of leap seconds that had occurred at the GPS epoch).  The definition says they steer to the defined epoch in UTC minus leap seconds, but it is the same either way (and is simpler to treat it as a constant from TAI, historical leap seconds are obviously never going to change).  Also better, because then if you aren't using UTC, you don't even need to _care_ about leap seconds.

If we look a little closer at http://tycho.usno.navy.mil/gpstt.html, they essentially state GPS-TAI = 19 +/-100ns is when SA is turned on.  When SA is turned off, they report 2-20ns rms.  I don't think we want to _model_ this noise, bu do make users aware that it is there, is circumstantial, and treat it as another base with fixed integer TAI+19 second offset, since that is what they steer it to.

I realize the current implementation essentially achieves this with the 19 second epoch offset, but it would be more ideal to use it like a base (since it is).  The current implementation adds confusion to the topic.

Really would be good to have format support for formats using gps_cycle, gps_week, second_in _week, optionally omitting gps_cycle too, in addition to seconds since GPS epoch.  Implementations using only 10-bit GPS week number still have a need to track cycles.

**Comment by mhvk** [comment]
@nikkelj - thanks for the nice summary and for also including links to actual precision, etc. Unfortunately, it seems that the issue describing in more detail how we would need to implement it (#6468) has stalled, even though there is quite a bit of code to start with. So, sadly it seems that for none of us it has risen high enough to actually tackle it, and we're basically waiting for someone to take the lead.... (hint, hint... ;-)

**Comment by nikkelj** [comment]
Roger that, this read like some folks may have been looking at it, but I am happy to take a look, and will do a PR if I solve it.

**Comment by mhvk** [comment]
Great! Do feel free to ask if you get confused.

**Comment by aarchiba** [comment]
I worked through this in detail for pulsar timing, and the details are in #2134, but there are unfortunately two different notions of GPS time scale - the GPS Combined Clock, or the Combined Clock corrected using almanac predictions of how best to line it up with UTC. The corrections associated with the two are about as different from each other as they are from zero.
